### ISSUE TITLE
Difference when using (defapi ...) and (api ...)

### ISSUE THREAD

**Issue by dpassen1** [metosin/compojure-api#159]
When using compojure-api 0.22.2 (and 0.22.1), I expect no difference between the following applications:

```clojure
(ns foobar.handler
  (:require [compojure.api.sweet :refer :all]
            [compojure.route :as route]
            [ring.middleware.defaults :refer [wrap-defaults site-defaults]]
            [schema.core :refer [defschema] :as schema])
  (:import [clojure.lang Keyword]))

(defschema Foo
  [schema/Keyword])

(defschema Bar
  [Keyword])

(defapi app-routes
  (swagger-docs {:info {:title "FooBar"}})
  (GET* "/foo" []
        :return Foo
        (constantly [:foo :is :great]))
  (GET* "/bar" []
        :return Bar
        (constantly [:bar :is :better]))
  (route/not-found "Not Found"))

(def app
  (wrap-defaults app-routes site-defaults))
```
and
```clojure
(ns foobar.handler
  (:require [compojure.api.sweet :refer :all]
            [compojure.route :as route]
            [ring.middleware.defaults :refer [wrap-defaults site-defaults]]
            [schema.core :refer [defschema] :as schema])
  (:import [clojure.lang Keyword]))

(defschema Foo
  [schema/Keyword])

(defschema Bar
  [Keyword])

(defn app-routes
  []
  (api
   (swagger-docs {:info {:title "FooBar"}})
   (GET* "/foo" []
         :return Foo
         (constantly [:foo :is :great]))
   (GET* "/bar" []
         :return Bar
         (constantly [:bar :is :better]))
   (route/not-found "Not Found")))

(def app
  (wrap-defaults (app-routes) site-defaults))
```

However, when running them and inspecting the generated swagger.json, the second version, not using defapi, contains this definition for Foo:
```JSON
"schema": {
    "type": "array",
    "items": null
},
"description": ""
```
when items should be of type "string".

For now, our workaround is to use clojure.lang.Keyword which works within defapi or within a plain api call.

**Comment by ikitommi** [comment]
That's strange. They should produce the same outputs. Would you have time to create a minimalistic sample project with `lein new compojure-api something` with the codes? 

Btw, both schemas are "primitive arrays", which is, for some reason, not implemented by ring-swagger. There is a ongoing PR to fix that.

https://github.com/metosin/ring-swagger/pull/56

**Comment by dpassen1** [comment]
Sure. [Here](https://github.com/dpassen1/foobar) is a small project generated from the compojure-api template.

The defapi branch works as expected.
The api-within-fn branch shows the issue.

**Comment by dparis** [comment]
Just bumped into this myself. Spent a frustrating amount of time trying to figure out why things weren't working like the example documentation until I tried using `(defapi ...)` instead of `(defn foo [] (api ...))`. 

For example, the app would throw an exception about a missing ctor when using a predicate schema to validate `:body` parameters when using `api` but it seems to work fine (or at least it compiles without exception) when using `defapi`. I was also seeing some validation errors originating in `ring-swagger` when passing the `{:coercion ...}` options map to `api`.

If I have time in the next few days, I'll try to put together a more thorough test case.

**Comment by ikitommi** [comment]
sorry for not having time on this. We finally got the Schema 1.0+ support and [Kekkonen](http://kekkonen.io) is out, will check this later today/tomorrow.

**Comment by ikitommi** [comment]
Ok, have a failing test for this in https://github.com/metosin/compojure-api/blob/defapi-vs-api-schemas/test/compojure/api/integration_test.clj#L1211-L1212. Have to figure out what's happening. `defapi` works as expected.

**Comment by dpassen1** [comment]
Thanks

**Comment by ikitommi** [comment]
Investigated this, and bumped into Clojure compiler oddities. First one is fixed in 1.8.0-alpha2. Will have to dig deeper with the other one, but have a idea on how to fix this (on the ring-swagger side).

**Comment by ikitommi** [comment]
should be fixed now in [`[0.24.1-SNAPSHOT]`](https://clojars.org/metosin/compojure-api/versions/0.24.1-SNAPSHOT)

**Comment by ikitommi** [comment]
`0.24.1` released.
